### PR TITLE
Added 3 fuzzers

### LINKED ISSUE #2927: Suggestion: Continuous Fuzzing
Hi, I'm Yevgeny Pats Founder of [Fuzzit](https://fuzzit.dev) - Continuous fuzzing as a service platform.

We have a free plan for OSS and I would be happy to contribute a PR if that's interesting.
The PR will include the following
- [go-fuzz](https://github.com/dvyukov/go-fuzz) fuzzers (This is generic step not-connected to fuzzit)
- Continuous Fuzzing of master branch which will generate new corpus and look for new crashes
- Regression on every PR that will run the fuzzers through all the generated corpus and fixed crashes from previous step. This will prevent new or old bugs from crippling into master.

You can see our basic example [fuzzitdev/example-go](https://github.com/fuzzitdev/example-go) and you can see an example of "in the wild" integration [google/syzkaller](https://github.com/google/syzkaller).

Let me know if this is something worth working on.

Cheers,
Yevgeny

### PR DESCRIPTION
**Pull request by AdamKorcz**
This PR answers https://github.com/gravitational/teleport/issues/2927 by adding 3 fuzzers to Teleport.

The fuzzers add no extra functionality to Teleport. They are tests themselves that run their targets with lots of pseudo-random data.

I suggest integrating Teleport into OSS-fuzz to allow the fuzzers to run continuously. I can confirm that these 3 fuzzers run on OSS-fuzz's infrastructure, and integration would mean that the fuzzers would run a couple of times per week. If a bug is found during those runs, detailed reports will be sent to the maintainers. OSS-fuzz is a free service that is offered with an implied expectation that bugs are fixed, so that the resources spent on fuzzing Teleport are put to good use.

I will be happy to integrate Teleport into OSS-fuzz. All I need are the email addresses for the maintainers to which potential bug reports should be sent. These email addresses will be publicly available and can be changed anytime.

I will be happy to work more on fuzzing Teleport in the future.

### CODE REVIEW

**Comment by awly** [comment]
ok to test

**Review by awly — COMMENTED** [review]
Thanks for the PR!
I agree that we could use some fuzzing. Will double-check with the team, in case anyone has objections.

In the meantime, could you move the `Fuzz` functions to standalone packages? Perhaps in subdirectories like `lib/fuzz/proxysubsysrequest`, `lib/fuzz/proxyjump` etc

Also, add a README in `lib/fuzz`, linking to OSS-fuzz and https://github.com/dvyukov/go-fuzz

**Review by webvictim — CHANGES_REQUESTED** [review]
As @awly said, thanks for the PR. We'd appreciate it if you could address his feedback.

**Comment by AdamKorcz** [comment]
Hi. Thank you for the feedback. I haven't forgotten about this or am ignoring it. I will get to it as soon as I can.

**Comment by webvictim** [comment]
No problem! Just marking that it needs changes so it’s out of our review queue for now :)

**Comment by AdamKorcz** [comment]
I have moved the fuzzers to a standalone package and have removed the fuzzer for the unexported target for now. 

Furthermore I have added a README with instructions on running the fuzzers locally.

File: lib/fuzz/README.md
Comment (awly) [inline lib/fuzz/README.md:8]: this is not needed

File: lib/fuzz/fuzz.go
Comment (awly) [inline lib/fuzz/fuzz.go:8]: `go-fuzz` requires the function to be named `Fuzz` exactly, without a suffix. This means that there's only 1 fuzzer per package/directory. These individual fuzzers must live in sub-directories of `lib/fuzz`.

If you want, we can merge this in and I'll move things around afterwards as needed.

**Comment by benarent** [comment]
@AdamKorcz PTAL at the code review.

**Comment by AdamKorcz** [comment]
@benarent Thanks for the ping. I will get to it this week!

File: lib/fuzz/README.md
Comment (AdamKorcz) [reply lib/fuzz/README.md]: This has been updated.

File: lib/fuzz/fuzz.go
Comment (AdamKorcz) [reply lib/fuzz/fuzz.go:8]: I would like to mention that the name of each harness is flexible if you opt to fuzz continuously through oss-fuzz. As a reference, check out [Prometheus' fuzzers](https://github.com/prometheus/prometheus/blob/master/promql/fuzz.go). These run fine through oss-fuzz. 
As mentioned earlier, I have the setup ready to integrate teleport into oss-fuzz, if you would like to proceed in that direction.
I am also fine with merging now and moving things afterwards.

Comment (awly) [inline lib/fuzz/fuzz.go:8]:
```suggestion
func FuzzParseProxyJump(data []byte) int {
```

File: lib/fuzz/fuzz.go
Comment (awly) [inline lib/fuzz/fuzz.go:16]:
```suggestion
func FuzzParseRoleVariable(data []byte) int {
```

File: lib/fuzz/fuzz.go
Comment (awly) [reply lib/fuzz/fuzz.go:8]: Ahh, I see. Please update `README.md` to point to `oss-fuzz` instead of `go-fuzz`, because you can't actually run these fuzzers locally with `go-fuzz`.

P.S. sorry for the response delay!

**Review by awly — APPROVED** [review]
@AdamKorcz sorry for the response delay!
LGTM.
Please rebase the PR and address the last few comments, I'll merge this PR afterwards.

**Comment by AdamKorcz** [comment]
I just noticed that `parse.RoleVariable` has changed to `NewExpression`. I will be updating this shortly.

**Review by awly — APPROVED** [review]

**Comment by awly** [comment]
Sorry for the long turnaround. LGTM.
@webvictim need your approval to unblock the merge.

**Review by webvictim — APPROVED** [review]

**Pull request merged**
